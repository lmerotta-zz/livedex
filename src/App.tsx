import React, { Suspense } from "react";
import { NetworkErrorBoundary, useResource } from "rest-hooks";
import InfiniteScroll from "react-infinite-scroll-component";
import useNextPage from "./utils/useNextPage";
import { CardContainer } from "./ui";
import SpeciesListResource from "./pokedex/resources/SpeciesList";
import SpeciesPlaceholder from "./pokedex/placeholders/SpeciesPlaceholder";

const SpeciesLazy = React.lazy(() => import("./pokedex/Species"));

const App: React.FC = () => {
  const list = useResource(SpeciesListResource.listShape(), {});
  const { fetcher } = useNextPage(SpeciesListResource.listShape());

  return (
    <NetworkErrorBoundary>
      <InfiniteScroll
        dataLength={list.results.length}
        next={fetcher}
        hasMore={list.next !== null}
        scrollThreshold="150px"
        loader={<SpeciesPlaceholder />}
      >
        <CardContainer>
          {list.results.map(r => (
            <Suspense key={r.name} fallback={<SpeciesPlaceholder />}>
              <SpeciesLazy name={r.name} />
            </Suspense>
          ))}
        </CardContainer>
      </InfiniteScroll>
    </NetworkErrorBoundary>
  );
};

export default App;
