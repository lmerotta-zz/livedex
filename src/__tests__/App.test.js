import React from "react";
import { render } from "@testing-library/react";
import { MockProvider } from "@rest-hooks/test";
import SpeciesListResource from "../pokedex/resources/SpeciesList";
import App from "../App";

jest.mock("../pokedex/Species", () => ({ name }) => <div>{name}</div>);

const fixtures = {
  loading: [],
  withInitialData: [
    {
      request: SpeciesListResource.listShape(),
      params: {},
      result: {
        count: 807,
        next: null,
        previous: null,
        results: [
          {
            name: "bulbasaur"
          }
        ]
      }
    }
  ]
};

describe("Species unit tests", () => {
  it("Renders the species when the resource is fetched", async () => {
    const tree = (
      <MockProvider results={fixtures.withInitialData}>
        <App />
      </MockProvider>
    );
    const { asFragment, findByText } = render(tree);
    const species = await findByText("bulbasaur");
    expect(species).toBeDefined();

    expect(asFragment()).toMatchSnapshot();
  });
});
