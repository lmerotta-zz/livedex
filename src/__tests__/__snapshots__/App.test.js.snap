// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Species unit tests Renders the species when the resource is fetched 1`] = `
<DocumentFragment>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <div
        class="flex flex-wrap"
      >
        <div>
          bulbasaur
        </div>
      </div>
      <div
        class="w-1/2 sm:w-1/3 md:w-1/4 lg:w-1/5 xl:w-1/6 px-2 my-2"
      >
        <div
          class="flex flex-col shadow-md rounded px-4 py-2 hover:shadow-xl"
        >
          <div
            class="flex flex-col"
          >
            <div
              class="h-24 w-24 self-center mb-2"
            >
              <span>
                <span
                  class="css-19bon8n react-loading-skeleton"
                  style="animation: animation-16jpnkj 1.2s ease-in-out infinite; width: 96px; height: 96px;"
                >
                  ‌
                </span>
              </span>
            </div>
            <h4
              class="text-sm leading-normal break-all"
            >
              <span>
                <span
                  class="css-19bon8n react-loading-skeleton"
                  style="animation: animation-16jpnkj 1.2s ease-in-out infinite; width: 40%;"
                >
                  ‌
                </span>
              </span>
            </h4>
            <p
              class="text-xs md:text-sm text-gray-400 leading-none"
            >
              <span>
                <span
                  class="css-19bon8n react-loading-skeleton"
                  style="animation: animation-16jpnkj 1.2s ease-in-out infinite; width: 50%;"
                >
                  ‌
                </span>
              </span>
            </p>
            <div
              class="my-3"
            >
              <span>
                <span
                  class="css-19bon8n react-loading-skeleton"
                  style="animation: animation-16jpnkj 1.2s ease-in-out infinite; width: 30%;"
                >
                  ‌
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
